### PR TITLE
Optionally run e2e pod as privileged for SELinux

### PR DESCRIPTION
**Pull request by bertinatto**
This PR adds back the SecurityContext field missed during the consolidation of redundant code in https://github.com/kubernetes/kubernetes/pull/79730. 

Also, it only runs the pod as privileged if the system has SELinux enabled and the volume isn't block (due to https://github.com/moby/moby/issues/35991).

**What type of PR is this?**

/kind flake

**What this PR does / why we need it**:

This is necessary for running some e2e tests on SELinux systems.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

/sig storage
CC @jsafrane @msau42

### CODE REVIEW

File: test/e2e/framework/volume/fixtures.go [hunk 1]
```diff
@@ -564,7 +565,7 @@ func testVolumeContent(client clientset.Interface, pod *v1.Pod, fsGroup *int64,
 // Multiple Tests can be specified to mount multiple volumes to a single
 // pod.
 func TestVolumeClient(client clientset.Interface, config TestConfig, fsGroup *int64, fsType string, tests []Test) {
-	clientPod, err := runVolumeTesterPod(client, config, "client", fsGroup, tests)
+	clientPod, err := runVolumeTesterPod(client, config, "client", fsGroup, false, tests)
```
Comment (msau42) [inline test/e2e/framework/volume/fixtures.go]: How come this one is false and the other is true?

Can we also make privileged a configurable parameter? I would like to be able to have most e2es not use privileged pods if the system doesn't support SELinux: https://github.com/kubernetes/kubernetes/issues/80186

**Comment by bertinatto** [comment]
Two jobs are failing because of a [docker issue](https://github.com/moby/moby/issues/35991) that prevents a privileged pod from mapping devices:

```
$ docker versioni

Client:
 Version:           18.09.9
 API version:       1.39
 Go version:        go1.11.13
 Git commit:        039a7df9ba
 Built:             Wed Sep  4 16:51:52 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          18.09.9
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.11.13
  Git commit:       039a7df
  Built:            Wed Sep  4 16:22:32 2019
  OS/Arch:          linux/amd64
  Experimental:     false

$ docker run -w /opt  --rm -it --device=/dev/console:/opt/10 busybox /bin/sh -c 'ls -l /opt/'
total 0
crw-------    1 root     root        5,   1 Oct 11 12:27 10
 
$ docker run -w /opt  --privileged --rm -it --device=/dev/console:/opt/10 busybox /bin/sh -c 'ls -l /opt/'
total 0
```

File: test/e2e/0 [hunk 2]
```diff
@@ -0,0 +1 @@
+
```
Comment (jsafrane) [inline test/e2e/0]: This file should not be part of this PR

File: test/e2e/framework/volume/fixtures.go [hunk 3]
```diff
@@ -488,6 +489,13 @@ func runVolumeTesterPod(client clientset.Interface, config TestConfig, podSuffix
 
 	for i, test := range tests {
 		volumeName := fmt.Sprintf("%s-%s-%d", config.Prefix, "volume", i)
+
+		privileged := false
+		if selinux.SELinuxEnabled() && test.Mode != v1.PersistentVolumeBlock {
```
Comment (jsafrane) [inline test/e2e/framework/volume/fixtures.go]: please comment why this is necessary - 1. selinux, 2. moby issue (+link)

Comment (bertinatto) [reply test/e2e/framework/volume/fixtures.go]: Added comment

File: test/e2e/framework/volume/fixtures.go [hunk 4]
```diff
@@ -488,6 +489,18 @@ func runVolumeTesterPod(client clientset.Interface, config TestConfig, podSuffix
...
 	for i, test := range tests {
 		volumeName := fmt.Sprintf("%s-%s-%d", config.Prefix, "volume", i)
+
+		// We make the container privileged when SELinux is enabled on the host,
+		// so the test can write data to a location like /tmp. Also, due to the
+		// Docker bug below, it's not currently possible to map a device with
+		// a privileged container, so we don't go privileged for block volumes.
+		// https://github.com/moby/moby/issues/35991
+		privileged := false
+		if selinux.SELinuxEnabled() && test.Mode != v1.PersistentVolumeBlock {
```
Comment (bertinatto) [inline test/e2e/framework/volume/fixtures.go]: This needs to be replaced by something else because it doesn't detect if SELinux is enabled on the host

Comment (alejandrox1) [reply test/e2e/framework/volume/fixtures.go]: @bertinatto could you please create an issue for this action item ?

**Review by alejandrox1 — COMMENTED** [review]
/priority important-soon
/milestone v1.17

**Comment by msau42** [comment]
/assign @jingxu97 
to check impact on windows tests

File: test/e2e/framework/volume/fixtures.go (hunk 1 above)
Comment (bertinatto) [reply test/e2e/framework/volume/fixtures.go]: > Can we also make privileged a configurable parameter?

So I figured this is a bit tricky to solve. These are the possible solutions that I thought about and their problems:

1. Make pods privileged:
	* We don't want this, specially where they are not required (on non-SELinux systems)
2. Make pods unprivileged and add label `svirt_sandbox_file_t` to node dirs where pods manipulate data (currently it's `/tmp` for hostPath)
	* Need to manually `chcon -R -t svirt_sandbox_file_t /dir_where_container_writes_files` in the host where test runs
3. Detect if node, where test is running, has SELinux enabled and make the pod privileged
	* I attempted this one, but it's tricky because we don't always know the node where the pod is going to be scheduled (so we could send `getenforcea command there and 
4. Have an env variable that toggles privileged
	* Not sure if this is a good idea, it needs to be toggled manually
        * It's not node-specific

Are there other alternatives?

File: test/e2e/framework/volume/fixtures.go (hunk 4 above)
Comment (bertinatto) [reply test/e2e/framework/volume/fixtures.go]: Sure, I'll create an issue as soon as I figure this out

File: test/e2e/framework/volume/fixtures.go [hunk 5]
```diff
@@ -488,6 +488,17 @@ func runVolumeTesterPod(client clientset.Interface, config TestConfig, podSuffix
 
 	for i, test := range tests {
 		volumeName := fmt.Sprintf("%s-%s-%d", config.Prefix, "volume", i)
+
+		// We need to make the container privileged when SELinux is enabled on the
+		// host,  so the test can write data to a location like /tmp. Also, due to
+		// the Docker bug below, it's not currently possible to map a device with
+		// a privileged container, so we don't go privileged for block volumes.
```
Comment (jingxu97) [inline test/e2e/framework/volume/fixtures.go]: Does mean for block volume, we cannot set SELinux enabled?

File: test/e2e/framework/volume/fixtures.go [hunk 6]
```diff
@@ -577,7 +588,7 @@ func TestVolumeClient(client clientset.Interface, config TestConfig, fsGroup *in
 // starting and auxiliary pod which writes the file there.
 // The volume must be writable.
 func InjectContent(client clientset.Interface, config TestConfig, fsGroup *int64, fsType string, tests []Test) {
-	injectorPod, err := runVolumeTesterPod(client, config, "injector", fsGroup, tests)
+	injectorPod, err := runVolumeTesterPod(client, config, "injector", true, fsGroup, tests)
```
Comment (jingxu97) [inline test/e2e/framework/volume/fixtures.go]: I think we cannot set this to true for all cases. For example, for windows, we cannot run privileged container.

File: test/e2e/framework/volume/fixtures.go (hunk 5 above)
Comment (bertinatto) [reply test/e2e/framework/volume/fixtures.go]: Actually, for block volume, we cannot make the *container* privileged. Specially if the container runtime is docker (due to the bug below).

File: test/e2e/framework/volume/fixtures.go (hunk 6 above)
Comment (bertinatto) [reply test/e2e/framework/volume/fixtures.go]: Added a check for Windows

**Review by bertinatto — COMMENTED** [review]
@jingxu97, PTAL

File: test/e2e/framework/volume/fixtures.go (hunk 5 above)
Comment (msau42) [reply test/e2e/framework/volume/fixtures.go]: Is there another directory we can safely write to that does not require privileged?

**Comment by kikisdeliveryservice** [comment]
Bug Triage Release Team checking in. Code freeze for the 1.17 release is on November 18. Is this still intended for 1.17? Thanks!

**Comment by timothysc** [comment]
/assign @bclau @timothysc 

@bclau - could you verify that this will work for windows use cases?

File: test/e2e/framework/volume/fixtures.go (hunk 5 above)
Comment (bertinatto) [reply test/e2e/framework/volume/fixtures.go]: > Is there another directory we can safely write to that does not require privileged?

Not that I'm aware of. 

By default, containers can only RW to files/directories labeled `svirt_sandbox_file_t`. So if a process breaks out of the container, it won't be able to change the host FS.

We *could* add the label above to a pre-defined directory, say */tmp/k8s-e2e*, and change the test to read/write data there, rather than the default */tmp*. See my comment https://github.com/kubernetes/kubernetes/pull/83727#discussion_r334856930. I've tried this and it worked.

Comment (bertinatto) [reply test/e2e/framework/volume/fixtures.go]: The downside of the approach above is that we need to change the test bootstrap to do `chcon -R -t svirt_sandbox_file_t /tmp/k8s-e2e` in the host before running the test.

I don't know if we are able to do that, however, this is IMO the right approach to solve our SELinux issues in e2e tests.

Comment (msau42) [reply test/e2e/framework/volume/fixtures.go]: I would prefer we look into this approach of prepping a special test directory first if the environment uses SELinux. That way we can remove the privileged setting on many tests (and maybe be able to add them to the conformance suite).  But this is sufficient for now.

**Review by msau42 — COMMENTED** [review]
/lgtm
/approve

File: test/e2e/framework/volume/fixtures.go (hunk 5 above)
Comment (bertinatto) [reply test/e2e/framework/volume/fixtures.go]: Just to clarify: we can remove the privileged setting if we prep a special test directory on SELinux environments. All containers would run as *unprivileged*.

FYI, I created the issue https://github.com/kubernetes/kubernetes/issues/84585 to track this.
